macros.py

```diff
@@ -164,6 +164,48 @@ def __repr__(self):
         return self.__str__()
 
 
+class GraphLine:
+    """An edge on the graph, representing one or more vulnerabilities"""
+    def __init__(self, origin, dest, label, colour):
+        self.origin = origin
+        self.dest = dest
+        self.label = label
+        self.colour = colour
+
+    def get_string(self):
+        """Convert the GraphLine into a line of a DOT file"""
+        return '"{origin}" -> "{dest}" [label="{label}" color="{colour}"];\n'.format(origin=self.origin, dest=self.dest, label=self.label, colour=self.colour)
+
+    @staticmethod
+    def merge_colours(lines):
+        """Take a set of edges to be merged and identify what colour the merged line should have"""
+        # Black (non-vendor specific) takes priority over red (vendor-specific)
+        for line in lines:
+            if 'black' == line.colour:
+                return 'black'
+        return 'red'
+
+    @staticmethod
+    def merge_set(lines):
+        """Identifies duplicate edges in a set of lines"""
+        sets = defaultdict(lambda: defaultdict(list))
+        for line in lines:
+            sets[line.origin][line.dest].append(line)
+
+        result = []
+        for origin, dest_set in sets.items():
+            for dest, line_set in dest_set.items():
+                if len(line_set) == 1:
+                    result.append(line_set[0])
+                else:
+                    # Alternative layout where there are 2 vulnerabilities
```

Comment:
2 or more

Reply:
Badly written comment - was meaning to say that the commented-out line below was an alternative layout to the one in use (i.e. for 2 vulnerabilities, show both names combined, and for 3 or more, just show the number). This didn't end up being very clear though, so I've just gone for showing the number any time there's more than one.